### ISSUE TITLE
New lint: Prefer mem::take over Vec::split_off(0)

### ISSUE THREAD

**Issue by lzutao** [rust-lang/rust-clippy#4953]
```rust
let mut vec = vec![1,2,3];
let vec2 = vec.split_off(0);
```

Should be written as

```rust
let vec2  = std::mem::take(&mut vec)
```

**Comment by lzutao** [comment]
Oh, wait. The semantics between tow functions are different.
Vec::split_off leave the caller with same capacity while mem::take leave call with default value
(in this case an empty Vec).

**Comment by Areredify** [comment]
I think `vec.drain(..).collect()` better represents the intention behind `vec.split_off(0)` (assuming it doesn't change capacity, which I don't think it is but seems like it's not specified either).

**Comment by Zalathar** [comment]
Note that the current implementation of `split_off` has special handling for the `0` case, which avoids the overhead of a memcpy but also results in the returned vector potentially having much more capacity than necessary.

So code that deliberately uses `split_off(0)` instead of `mem::take` might not actually be managing capacity in the way the author expects.

See https://github.com/rust-lang/rust/issues/119913 for more context.
